### ISSUE TITLE
Map dragging is slow when using more than 15 000 markers in a high zoom level

### ISSUE THREAD

**Issue by JonasDaWi** [googlemaps/android-maps-utils#29]
Hey,
I have (sometimes) much more than 15 000 markers. When i drag the map in a high zoom level e.g. above an area that has only like 200 (clustered) markers it is very slow although I have a Galaxy S3. Someone knows how to make dragging fast in high zoom levels?

**Comment by Calius** [comment]
Is there any practicable solution for this issue yet?

**Comment by newmanw** [comment]
Any word on this?  I have about 8k markers and the lib is unusable at that point unfortunately.  And I say unfortunately with great respect as this is a beautiful lib as far as UI, very easy to use and documented well.  Really hoping I can use this soon :( 

At a high level zoom things are pretty good.  However as I start to zoom in performance degrades.  I assume this is due to markers and clusters off the viewport being drawn/redrawn as I zoom.  

Then once I am zoomed all the way in if I zoom out quickly I notice a huge hit.  All the markers are unclustered (as I was all the way zoomed in before).  I keep zooming out until I can see the entire map.  I can see all (unclustered) markers and it takes 20-30 seconds to loop through all 8K markers and place them back into a clusters.

All of this probably relates to the fact that off screen markers and clusters are being drawn.  However I also wonder if this is slow due to the animations.  In cases like this it might make sense to be able to turn off animations and just remove and redraw all clusters and marker on zoom change.

**Comment by akirmse** [comment]
I made the following changes locally to get the library to work reasonably
well with lots of markers:

- Disable animation completely.  There is a constant you can change in the
existing code for this.

- When camera motion stops, use an AsyncTask to remove all items from the
ClusterManager, and then re-add only those visible on the screen.  (I do
this only if there's a large number of markers.)

- In ClusterManager, I imported the
mClusterTask.executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR) change from
upstream (I had an older version of the library).

- I added a cancel() method on Algorithm, and a periodic check in
NonHierarchicalDistanceBasedAlgorithm.getClusters, so that I could cancel
previous clustering operations when starting a new one.


I like this library but I wish the team were more responsive to obvious
problems like this and
https://code.google.com/p/gmaps-api-issues/issues/detail?id=5313

**Comment by newmanw** [comment]
@akirmse agreed, wish issues like this would get more attention.  Of course I am not contributing so I don't have much to complain about ;)

Is this being actively worked by contributors, or has anyone forked, maybe a possible PR for this?

**Comment by emkou** [comment]
For more than two years the clustering element of this library is totally unusable. Clustering more than 100 items even on a big area brings huge performance issues. This should be either adressed or removed completely from the library.

**Comment by stephenmeyer** [comment]
Heres my hacked version:
Update getClusters to 
> public Set<? extends Cluster<T>> getClusters(double zoom, LatLngBounds bounds)

Add the following if to the QuadItem<T> candidate for each loop
> if(!bounds.contains(candidate.mPosition)) {
> continue;
> }

Replace ClusterManager.cs onCameraChange function with the following:
>mMap.clear();
>mPreviousCameraPosition = mMap.getCameraPosition();
>currentBounds = mMap.getProjection().getVisibleRegion().latLngBounds;
>Set<? extends Cluster<T>> results = mAlgorithm.getClusters(mPreviousCameraPosition.zoom, currentBounds);
>for(Cluster<T> cluster : results)
>{
>   // Strip out DefaultClusterRendered.cs perform method and render each cluster accordingly here
>}

**Comment by gintechsystems** [comment]
@akirmse can you show your example in code?

**Comment by WonderCsabo** [comment]
@broady any update on this one? Your clustering looks very cool, the animations are great, but it just cannot handle more markers. And this is a big problem, since the point of clustering that we have lots of markers. I am currently using [this](https://github.com/zamesilyasa/android-maps-utils/commit/1cf9f4e553fded253b0fcb8d31f58640687de985) algorithm to only cluster the visible markers, but this also has drawbacks (clusters change a lot, user have to wait a little bit when cluster show up after changing camera).
Are you planning to implement robust solution? Thanks in advance for your answer, and for this great library in general! :)

**Comment by gintechsystems** [comment]
I had to modify this library to my needs to get it working. I turned off animation and have it refreshing async on camera change. I have about 50k+ markers and it works great. It will only show markers in the visible region.

**Comment by broady** [comment]
Hey folks.

I've actually left the Maps team, so everything I do on this project is in my spare time.

Here's what I think should happen:
* `DefaultClusterRenderer` implements `GoogleMap.OnCameraChangeListener`.
* Somehow pipe camera change events to `DefaultClusterRenderer`.
* On the first camera change event, set a flag that changes the renderer to only render markers inside the visible bounds.
* Update visible markers when the camera changes and/or clusters change.

Unfortunately the loose coupling design of the library isn't conducive to this kind of stuff. I think it was a mistake.

@jfschmakeit @markmcd in case they have spare cycles.

**Comment by newmanw** [comment]
Anyone on maps team supporting this or is it left to die?

**Comment by broady** [comment]
I've just written down the exact steps of my preferred solution. Please discuss, send a pull request, etc! :)

**Comment by WonderCsabo** [comment]
@broady it is sad you no longer work on this officially, and no one else took over. This is an important project, even referred from the official doc. Thanks for still managing the project, your time is much appreciated!

I already tried to implement your solution, but only in a very naive way. `ClusterManager` automatically calls `mRenderer.onCameraChange` if the renderer implements `OnCameraChangeListener`, so piping that is already done. This is my `onCameraChange` method in the renderer:

```java
@Override
public void onCameraChange(CameraPosition cameraPosition) {
  LatLngBounds bounds = map.getProjection().getVisibleRegion().latLngBounds;

  for (Marker marker : clusterManager.getClusterMarkerCollection().getMarkers()) {
      if (bounds.contains(marker.getPosition())) {
          marker.setVisible(true);
      } else {
          marker.setVisible(false);
      }
  }
}
```

There are two problems with this approach:
* when panning/rotating, the movements are slow because this loop blocks the drawing thread for too much time by iterating over all the clusters and checking they are in the bounds or not
* this does not help at all when zooming, because the markers are added asynchronously that time

So a more complicated solution is needed, but i could not figure that out.

**Comment by meierjan** [comment]
+1 for this issue

**Comment by arnaudbos** [comment]
Lol, French people are famous for complaining a lot so keep this in mind, plus I'm not contributing either, but for a library that is referenced on the official documentation I see it as a "fuck you" from the Maps team to not support this at this time.
I mean, look at the performances of the clustering libraries for iOS... Any random app can have blazingly fast map with hundred of points.

**Comment by bbreukelen** [comment]
With help of stephenmeyer's hack I implemented my version of viewport only clustering which is lightning-fast.

Here is a detailed description of all my changes.
Because my time was limited and I had to reuse the viewport bounds I created a tiny class to store the value in. Quick & dirty but it doesn't hurt anyone.

So, create a class in android/clustering. I called mine bvb.java and this is the content:

```
package com.google.maps.android.clustering;
import com.google.android.gms.maps.model.LatLngBounds;

public class bvb {
    public static LatLngBounds mapClusterViewportBounds;
    public static boolean infoWindowShowing = false;
}
```
This holds 2 variables that can be accessed anywhere within the clusterManager.
The mapClusterViewportBounds variable will contain the coordinates of the part in the map that is visible to the user.
The infoWindowShowing variable should be set to true if you are using an infowindow popup when a user clicks on a marker. In the usual code, the clustering is only recalculated when the zoomlevel changes. If a user clicks on a marker, it's centered and the map doesn't zoom. The infowindow shows and stays in place. But with viewport clustering implemented, panning should also recalculate clusters or else you'd not see anything until you zoom. If a marker is now clicked, the infowindow shows, the marker is centered which now triggers reclustering and the infowindow is removed again.
To prevent this, whenever you are showing an infoWindow, add bvb.infoWindowShowing = true;
When an infoWindow is closed, add bvb.infoWindowShowing = false;
Note that you should also add bvb.infoWindowShowing = false; whenever you load new data on the map as the infoWindow is also removed when the map is cleared.


Now edit ClusterManager.java
In the import section add the following:
```
import android.util.Log;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.LatLngBounds;
```

Find the onCameraChange method and change it to the following:
```
    @Override
    public void onCameraChange(CameraPosition cameraPosition) {
        // Added this to store the viewport bounds and increase slightly for user convenience
        LatLngBounds b = mMap.getProjection().getVisibleRegion().latLngBounds;
        LatLng bNE = b.northeast;
        LatLng bSW = b.southwest;
        double bAdjustLat = (bNE.latitude - bSW.latitude) * 0.15; // 15% increase
        double bAdjustLon = (bNE.longitude - bSW.longitude) * 0.15; // 15% increase
        bNE = new LatLng(bNE.latitude + bAdjustLat, bNE.longitude + bAdjustLon);
        bSW = new LatLng(bSW.latitude - bAdjustLat, bSW.longitude - bAdjustLon);
        bvb.mapClusterViewportBounds = new LatLngBounds(bSW, bNE);

        if (mRenderer instanceof GoogleMap.OnCameraChangeListener) {
            ((GoogleMap.OnCameraChangeListener) mRenderer).onCameraChange(cameraPosition);
        }

        if (bvb.infoWindowShowing) {
            Log.i("Cluster", "Not reclustering as we have an infoWindow");
            return;
        }
        cluster();
    }
```
We first load the viewport bounds into a variable. I increased the viewport with 15% so that the corners also show markers/clusters and panning shows at least some data to the sides. Just 15% extra won't hurt performance. You can play with this number and increase/decrease it to your liking.
Then I check if there's an infoWindow set. If so, it will not recluster until it's closed and the onCameraChange method is called.


Now edit algo/NonHierarchicalDistanceBasedAlgorithm.java.
In the import section add the following:
```
import android.util.Log;
import com.google.maps.android.clustering.bvb;
```

In the getClusters method find the line "for (QuadItem<T> candidate : mItems) {
Underneath that is an if section about visitedCandidates.contains(candidate)).
Below that if-section, add this following code (for me it's on line 99).

```
if (bvb.mapClusterViewportBounds != null && !bvb.mapClusterViewportBounds.contains(candidate.mPosition)) {
                    continue; // Added to exclude markers outside viewport
}
```
While iterating over all map-points, this doesn't process any point outside your viewport + the 15%.
Which is what make the clustering perform quite well.


Then edit algo/PreCachingAlgorithmDecorator.java.
Find the getClustersInternal method.
Uncomment the following line:

`mCache.put(discreteZoom, results);`

The cluster-results are cached per zoomlevel. Since we now have viewports-clusters, this no longer works. I thought about caching by zoomlevel and viewport-coordinates, by the chance of hitting the exact same combination multiple times is slim and since it already performs pretty well I decided to disable caching completely. I didn't back-out the caching-code but went for the quick approach of just disabling saving to cache so it will never find any cache.

[Here](https://www.dropbox.com/s/qa6dmxionfpmobl/bvbClusterHack.tar?dl=1) is an archive with the changes files.

Bo

**Comment by Libby713** [comment]
As one solution to slow animation you can now choose to disable the animations, which was added in 27f7a5b

**Comment by CeccoCQ** [comment]
I've a map with 4k markers on it. I've tried the VisibleBasedAlgorithm but the performance remains very very slow. 
When I zoom out the map, I've to wait some seconds before the clustering ends.
When I zoom in, the cluster disappear very slow, so the user is confused.

My experience:
- I've disabled the animations;
- I've used the https://github.com/zamesilyasa/android-maps-utils (branch issue-82)
- I've added further optimization on ClusterRenderer class

Is there a way to avoid the cluster expansion when the markers aren't visible on maps bounds? I think that this could be only solution.

**Comment by Gary111** [comment]
@CeccoCQ, you can try this optimised variant of Screen Based algorith [https://github.com/zamesilyasa/android-maps-utils/commit/8694da38220d6394e4d9216b2a359d536afe9c40](https://github.com/zamesilyasa/android-maps-utils/commit/8694da38220d6394e4d9216b2a359d536afe9c40).

**Comment by CeccoCQ** [comment]
Hi @Gary111,
I've finish the experiments with your code just now and I've seen some improvements.

By your experience, is there a way to increase the bitmap drawing performance?

**Comment by zamesilyasa** [comment]
Are you sure you are  using as small bitmaps(thumbnails) for markers? if bitmap is large, it will slow rendering.

**Comment by Gary111** [comment]
@CeccoCQ, If you mean drawing markers, I use disk and memory LRU cache for this. So I generate icon only once and save it in cache, next time I just extract it form cache. It increases performance a little.

**Comment by CeccoCQ** [comment]
Hi, thanks for your responses.

This is my code used to create the bitmap:
```
public LruCache<Integer, BitmapDescriptor> mIcons;
.....
int sizeSelected = Math.round(TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 50, context.getResources().getDisplayMetrics()));
Bitmap selectedBitmap = BitmapFactory.decodeResource(context.getResources(), mso.markerSelected).copy(Bitmap.Config.ARGB_8888, true);
Bitmap selected = Bitmap.createScaledBitmap(selectedBitmap, sizeSelected, sizeSelected, true);
BitmapDescriptor descSelected = mIcons.get(mso.markerSelected);
if (descSelected == null) {
    descSelected = BitmapDescriptorFactory.fromBitmap(selected);
    mIcons.put(markerTypeId, descSelected);
}
```

then into the onBeforeClusterItemRendered method:
```
BitmapDescriptor desc = mIcons.get(markerTypeId);
```

The original bitmap is a png file of 294x294px of 15Kb

**Comment by PJHaynes304** [comment]
So... over 4 years down the line is there any progress on this issue?? 
I've just migrated from AME to AMU and the performance is MUCH MUCH worse. 
I thought it would be a step forwards in performance but it's been a huge leap backwards.
Even with animations turned off it takes an age for clusters to be regrouped when zooming out.

**Comment by PJHaynes304** [comment]
The fact this issue has been open for 5 years and the lack of response from anyone suggests this will not be addressed. I’d recommend a 3rd party clustering library if you want something that works.

**Comment by hannesa2** [comment]
I can confirm, this https://github.com/mg6maciej/android-maps-extensions/tree/develop can handle a lot of markers without performance issues

**Comment by PJHaynes304** [comment]
Having used android-maps-extensions and google-maps-clustering I would recommend the following:

https://github.com/sharewire/google-maps-clustering

Performance is fantastic and the code is well written and easily extended. Also has nice animations when the map moves and the clusters regroup.
